core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -211,7 +211,7 @@ abstract class RDD[T: ClassTag](
   // Our dependencies and partitions will be gotten by calling subclass's methods below, and will
   // be overwritten when we're checkpointed
   private var dependencies_ : Seq[Dependency[_]] = null
-  @transient private var partitions_ : Array[Partition] = null
+  @transient @volatile private var partitions_ : Array[Partition] = null
```

Comment:
per my understanding, I don't think we need the `@volatile` here, probably the only place we need to change is the add the modifier of `synchronized` for method `getPartitions` in the concrete sub class of RDD, which will force the cpu cache to memory as the barrier fence of jvm memory model.